### DISCUSSION TITLE
redis-cache:metrics key growing up infinitely

### DISCUSSION THREAD

**Discussion by kamilgrzegorczyk** [rhubarbgroup/.github#46] [Support (ROC)] [answered]
`redis-cache:metrics` key is growing over time

## Description
The issue happens only on one of dozens of sites I have redis installed on. It started happening over a year ago.
For whatever reason metrics key grows over the few days and gaining even 400-500megabytes.

The site has basic redis configuration

## Expected Behavior
Metrics consume max 5-10 megabytes of memory

## Actual Behavior
Metrics key consume 500 megabytes (see attached screen)

## Possible Fix
Disabling metrics for that particular site (dummy fix)

## Steps to Reproduce
I can't provide any steps for reproducing that bug

## Additional context
It just consumes a lot of memory which instead should be available for the "real data" keys

The site is running on two domains (completely separated, shared codebase but separate infra) and the issue exists on both.
Both of the domains are med size traffic (tens of thousands users per day)

## Environment
As the bug occurs since over a year therefore it was happening under a range of versions up until the most recent
Plugin version - most probably since 2.3.x
WP version - most probably at least since 6.2.x
PHP version 8.0.x-8.1.x

Site is converted to a multisite since 2 months but the issue was happening for over a year when it was still a single site.

**Comment by kamilgrzegorczyk** [comment]
Forgotten screenshot
![Screenshot 2023-12-29 at 08 55 44](https://github.com/rhubarbgroup/redis-cache/assets/665560/ea86e209-3794-49ed-9865-4a11b762b173)

**Comment by tillkruss** [comment]
Please post your diagnostics.

**Comment by kamilgrzegorczyk** [comment]
Here you are

```
Status: Connected
Client: PhpRedis (v6.0.2)
Drop-in: Valid
Disabled: No
Ping: 1
Errors: []
PhpRedis: 6.0.2
Relay: Not loaded
Predis: 2.1.2
Credis: Not loaded
PHP Version: 8.1.26
Plugin Version: 2.5.0
Redis Version: 7.0.7
Multisite: Yes
Metrics: Enabled
Metrics recorded: 982398
Filesystem: File modifications are not allowed.
Global Prefix: ""
Blog Prefix: 1
WP_REDIS_CLIENT: "phpredis"
WP_REDIS_HOST: "redacted-out"
WP_REDIS_PORT: 6379
WP_REDIS_DATABASE: 1
WP_REDIS_TIMEOUT: 1
WP_REDIS_READ_TIMEOUT: 1
WP_REDIS_PREFIX: "bnd_"
WP_REDIS_PLUGIN_PATH: "/var/redacted-out/"
Global Groups: [
    "blog-details",
    "blog-id-cache",
    "blog-lookup",
    "global-posts",
    "networks",
    "rss",
    "sites",
    "site-details",
    "site-lookup",
    "site-options",
    "site-transient",
    "users",
    "useremail",
    "userlogins",
    "usermeta",
    "user_meta",
    "userslugs",
    "redis-cache",
    "blog_meta",
    "network-queries",
    "site-queries",
    "user-queries"
]
Ignored Groups: [
    "counts",
    "plugins",
    "theme_json",
    "themes"
]
Unflushable Groups: []
Groups Types: {
    "blog-details": "global",
    "blog-id-cache": "global",
    "blog-lookup": "global",
    "global-posts": "global",
    "networks": "global",
    "rss": "global",
    "sites": "global",
    "site-details": "global",
    "site-lookup": "global",
    "site-options": "global",
    "site-transient": "global",
    "users": "global",
    "useremail": "global",
    "userlogins": "global",
    "usermeta": "global",
    "user_meta": "global",
    "userslugs": "global",
    "redis-cache": "global",
    "blog_meta": "global",
    "network-queries": "global",
    "site-queries": "global",
    "user-queries": "global",
    "counts": "ignored",
    "plugins": "ignored",
    "theme_json": "ignored",
    "themes": "ignored"
}
Drop-ins: [
    "Redis Object Cache Drop-In v2.5.0 by Till Krüss"
]

```

**Comment by tillkruss** [comment]
The config looks good. You have 982,398 metrics recorded. If it's taking up too much space, use the `WP_REDIS_DISABLE_METRICS` constant to disable them entirely.

**Comment by kamilgrzegorczyk** [comment]
Thank you @tillkruss 

Well the current situation is when the DB was pruned few days ago so that doesn't reflect the state when its actually having 500mbs.

Do you know why the number of metrics can grow so high? What are the rules for pruning/truncating the logs?

What I observed is that the issue happens mostly in the times when the admin traffic is low (like current Christmas time). The environment is dockerized and the sites work in headless mode so admin per se is not hit. Do you prune/truncate metrics on certain admin actions or cron events? (To answer the not asked question - I have server side cron running).

Maybe I will just dig it up in the source code and see what could be causing the overflow in my scenario.

**Comment by tillkruss** [comment, marked as answer]
>Do you know why the number of metrics can grow so high? What are the rules for pruning/truncating the logs?

One HTTP request will case one metrics entry.

There is a `rediscache_discard_metrics` scheduled event that will delete all metrics older than `WP_REDIS_METRICS_MAX_TIME` (1 hour by default).

So if you WP Cron isn't working properly it may build up over time.
